### ISSUE TITLE
Integration to Volumio as plugin, system integrity check failed

### ISSUE THREAD

**Issue by misko903** [antiprism/mpd_oled#52]
Hello,
I am using this plugin since I my first Raspberry - 3B+ with volumio on it. I have installed display at the beginning too! I love your work.
The only problem is, I must reinstall it everytime after Volumio get update, as Volumio is checking system integrity.
Is it possible to make your plugin an official Volumio plugin to install it simply from repository?

Thank you,
Michal, Slovakia

**Comment by antiprism** [comment]
Hi Michal

I'll give a slightly longer answer, to explain the situation and suggest possible workarounds.

mpd_oled needs a copy of the playing audio to show the spectrum. The mpd_oled instructions change /volumio/app/plugins/music_service/mpd/mpd.conf.tmpl, which is a Volumio system file, and I imagine that it is just this change that is causing the system integrity check to fail (I don't think it will care about changes in the home directory or new binaries installed, but I haven't checked). In which case, if you keep a copy of the original mpd.oled.tmpl file, and copy it back before an upgrade, that might be enough.

Looking at the Volumio docs https://volumio.github.io/docs/API/Command_Line_Client.html you may be able to run the following command to remove the custom change
```
updater restorevolumio
```

Similarly, if you use the ALSA method to make a copy of the audio ( https://github.com/antiprism/mpd_oled/issues/4 ), rather than changing the Volumio mpd.conf.tmpl file, there should be no sytem changes  and the upgrade should hopefully work.

I have seen messages on the Volumio forum that suggest it will be easier to make a reliable copy of the audio in the future. I'll review the feature it it becomes available as it should also avoid the integrity check issue.

Regarding creating a plugin, this isn't something I would do. However, I am currently reworking the mpd_oled build so that it is easy to make a binariy that will work on all Pis and supported Player OSs. This will make installation easier, but also makes the job of creating a plugin easier if soeone wishes to do this.

I think it would be useful to keep this issue open. Please report anything that helps with the upgrade issue and I can add it to the installation instructions.

Adrian.

**Comment by supercrab** [comment]
I wrote a volumio plugin a couple of years ago and have thought about writing one for `mpd_oled`.  It's on my list of things to do though and I think it would be useful for configuring `mpd_oled` once it has been built by the user.

A volumio plugin is nothing more than a javascript file that can respond to various volumio events.  They save their config in a corresponding JSON file and can run OS commands (I'm sure I saw one that did).  So in theory, if the plugin in theory should be able to adjust the `ExecStart` line in the `mpd_oled.service` file and run `install.sh` to install the changes  (assuming the plugin process has rights to access the folder and execute the script)

Will report back when/if I get a chance to experiment :)

**Comment by antiprism** [comment]
Hi Mase

Sounds good, although it might be some work to support all the mpd_oled options in a UI!

Adrian.

**Comment by supercrab** [comment]
@antiprism oh, the UI front end of a plugin is defined by a JSON file, so if you know all the available options, it's straight forward enough to do

[https://github.com/volumio/volumio-plugins/blob/master/plugins/system_controller/gpio_control/UIConfig.json](example UI config from my plugin)

**Comment by misko903** [comment]
Hi Adrian, hi all,
thanks for longer answer, you explained a lot. Next time I will try to use the ALSA method or even I will try to use `updater restorevolumio` . There is an option in Volumio, where can I restore the system, but It will delete all the custom settings as connected drives, favourite radios... Practically it is clearly installed system. Then I am able to update and the procedures of setting up begins.

One option, instead of 
`sudo nano /boot/config.txt` It is possible to use `sudo nano /boot/userconfig.txt` , which will stay unchanged after restoring the system.

So I'll keep it is for now, but I'll try it on next Volumio upgrade.
Thanks
Michal

@supercrab , this would be awesome!

**Comment by antiprism** [comment]
Hi Michal

Thanks for the suggestion. You may have some old instructions for installing mpd_oled on Volumio as the current instructions now include commands for editing   /boot/userconfig.txt

Adrian.

**Comment by antiprism** [comment]
Hi Mase

I backported some of my development code into the current mpd_oled and then, after several hours of tweaking build options, was finally able to produce a single mpd_oled binary, on a Pi 4 running Raspbian buster, that will work on both Volumio and Moode on a Pi Zero (I haven't checked RuneAudio, see below).

The build takes a bit longer, especially on a Pi Zero, but I imagine most people will choose a binary install if it is available.

I'll complete this development (I haven't updated the documentation yet), and see if I can write a script to produce a full binary installation (that also includes cava) as a deb file. Hopefully a plugin could then just install this deb file during installation.

Adrian.

Note for RuneAudio: the downloadable version is very old and I don't think it works on a Pi Zero, and people seem to run a number of newer custom images shared through the forum, but it is often hard to install a working compiler and so a binary install of mpd_oled would be especially useful there.

**Comment by supercrab** [comment]
@antiprism 

Nice work!  I just checked the documentation of [Volumio plugins](https://volumio.github.io/docs/Plugin_System/Writing_A_Plugin.html) and it does support an install bash script, so that should be able to do the installation.

Here's an example of a plugin install script for illustrative purposes.  It looks like it could be used to clone and compile the source, without the need for binaries or it could do pretty much anything it needs to.

```#!/bin/bash

echo "Installing Spop Dependencies"
sudo apt-get update
sudo apt-get -y install libasound2-dev libreadline-dev libsox-dev libsoup2.4-dev libsoup2.4-1 libdbus-glib-1-dev libnotify-dev --no-install-recommends



echo "Installing Spop and libspotify"

DPKG_ARCH=`dpkg --print-architecture`

echo $DPKG_ARCH
cd /tmp
wget http://repo.volumio.org/Packages/Spop/spop-${DPKG_ARCH}.tar.gz
sudo tar xvf /tmp/spop-${DPKG_ARCH}.tar.gz -C /
rm /tmp/spop-${DPKG_ARCH}.tar.gz


sudo chmod 777 /etc/spopd.conf

#required to end the plugin install
echo "plugininstallend"```

**Comment by antiprism** [comment]
Hi Michal

I can confirm that Volumio can be updated if the following command is run first (with the understanding that it may also undo other custom changes not related to mpd_oled)
```
volumio updater restorevolumio
```
All I had to do after the restart was reconfigure the audio copy
https://github.com/antiprism/mpd_oled/blob/master/INSTALL_VOLUMIO.md#configure-copy-of-audio

I'll add this to the instructions.

Adrian.

**Comment by misko903** [comment]
Thanks for the info! When I was doing the restore through volumio GUI, it
has deleted all my settings incl. All the radio stations, music sources and
favourites.

I'll try it on next update, it will save me a lot of time :)

I am watching also the work around the official plugin. Impressive!!!

Dňa pi 29. 1. 2021, 13:30 Adrian Rossiter <notifications@github.com>
napísal(a):

**Comment by antiprism** [comment]
Hi Mase

I have just pushed a development branch. You can get a copy with
```
git clone -b dev https://github.com/antiprism/mpd_oled.git
```
I will make a binary installation next, but there are installation changes to help with the binary installation that should also make life easier for you with the plugin.

The instructions now install mpd_oled before testing it, and the installation includes some scripts to install and manage the mpd_oled service. You can run `sudo mpd_oled_install.sh` to install the service (it uses a dummy OLED setting, which causes the mpd_oled command to fail, and the service simply does not run).

The mpd_oled options that the service will use are set by running `sudo mpd_oled_edit.sh` followed by the options, and the service file will be updated and the program started. I think this could also work for testing command line options, rather than running the mpd_oled command directly, because as soon as the options are good the configuration has been set and there is no need to do anything else. (It might be an idea for the script to print any mpd_oled error message.)

I will also script the change to mpd.conf.tmpl. I have asked on the Volumio forum if anyine knows a command to trigger the regenereration of mpd.conf, but have not yet received a reply.

Adrian.

Further note for RuneAudio: it looks like it has been forked, so I will see if I can support the fork: https://github.com/rern/rAudio-1/

**Comment by supercrab** [comment]
Thanks for the update!  I'll check it out in a bit :)

So far, I'm about 25% done with the plugin.  I had a few blockers to do with the way Volumio plugins are implemented and some things aren't documented but it should be straightforward from now.

 I've managed to get the plugin to start execution of `mpd_oled` which is promising.  I couldn't get it to run `install.sh`  to install the service because it needs `sudo` authority.  This is strange because I'm using the following line in JS to kill `mpd_oled` before starting it again `exec('/usr/bin/sudo /usr/bin/killall mpd_oled')` 

I haven't looked at automating the install of mpd_oled and configuring the copy of the audio yet...

Cheers
Mase

**Comment by antiprism** [comment]
Hi Mase

I have managed to get help on the Volumio forum for a script to trigger the regeneration of /etc/mpd.conf

   https://community.volumio.org/t/command-to-regenerate-mpd-conf/44573

That means I can now write a script which will add the audio_output section to /etc/mpd.conf without user intervention.

However, the system integrity check will fail even by running `touch` on a file, and so for the uninstall script I will see if it is possible to restore of just the mpd.conf.tmpl file (rather than all of /volumio).

Adrian.

**Comment by antiprism** [comment]
Hi Mase

I have attached a zip file with a couple of scripts for review. The install script adds the audio_output section to /etc/mpd.conf and the uninstall script removes it again. Run `cat /etc/mpd.conf` after running either script to see it is updated.

The uninstall script uses the technique used by "volumio updater restorevolumio" to just restore the mpd.conf.tmpl file. However, rather than rebooting it just remounts the root filesystem, and this seems to synchronise the overlay filesystem. I don't know enough to say if this is a good idea or not, but it does appear to work. It is probably better not to run it on your favorite Volumio setup just yet! I'll ask on the Volumio forum if someone could review the script.

Adrian.

**Comment by supercrab** [comment]
Ooh nice!  Looks good to me!

**Comment by antiprism** [comment]
Hi Mase

That is good progress with the plugin, and it will be multilingual too!

I haven't made any binary packages yet, but it is easy to build the binaries. If you look at the last two commands in commands.txt
https://github.com/antiprism/mpd_oled/blob/dev/commands.txt
They are commands to run when configuring mpd_oled and cava, and when you run `make install-strip` afterwards they make binaries that run on a Pi Zero and are installed under /home/volumio/inst. This is the install tree that I will be packaging.

Adrian

**Comment by misko903** [comment]
If you need Slovak, maybe Czech translation, I can help.

M

Dňa št 4. 2. 2021, 17:59 Adrian Rossiter <notifications@github.com>
napísal(a):

**Comment by antiprism** [comment]
I could do the Spanish translation.

**Comment by supercrab** [comment]
I could ask a friend to do a translation in Welsh :)  I don't have to make it multilingual but it looks neater and it's a more complete solution.  It is making the code more complex and is taking me some extra time to support it.   It's handy to have all the strings contained in the plugin contained in a separate JSON file, so any changes there will also appear in the UI :)

I did read the installation guide for installing in the end - I found it.  I don't know anything about install trees or much about Linux in general.  Fingers crossed I can get it working, the heat is on!

**Comment by antiprism** [comment]
Hi Mase

There is a mistake with quoting in the install script, the variable that sets the text to insert should look like
```
mpd_conf_txt="
# add a FIFO to be read by a cava, which is run as a subprocess of mpd_oled
audio_output {
        type            "\""fifo"\""
        name            "\""mpd_oled_FIFO"\""
        path            "\""/tmp/mpd_oled_fifo"\""
        format          "\""44100:16:2"\""
}"
```
Also, mpd_oled cannot be running when the install script is run. I need to review the conditions for the scripts to work and try to
warn or account for them. One of the problems seems to be whether mpd can open/close/create/delete the FIFO when cava also has it open.

Adrian.

**Comment by supercrab** [comment]
It should be possible to use `killall mpd_oled` and the same for mpd.  If MPD is killed how do we restart it?

I have some code tidying, finish up multi language support, finish off the UI JSON, little refactor here and there and then I’ll be onto install and uninstall malarkey. I managed to add a button to the UI which runs `i2cdetect` if it’s present (i need to add this to the install) and it tells you what address your oled screen is. I have to account for multiple devices on the bus on either bus.

**Comment by antiprism** [comment]
Hi Mase

MPD will be run as a service, so `systemctl stop mpd` should stop it, and then it can be started again with ` systemctl start mpd` (checking that Volumio doesn't also try to restart it, maybe run `mpc stop` first). mpd_oledcould be running as a service or on the command line. It is possible to check if it is running as a service with `systemctl is-active --quiet mpd_oled`, in which case it can be stopped and started again afterwards. And if it is not running as a service then run `killall mpd_oled` and don't start it again.
I'll add it like this to the install and uninstall scripts for the mpd.conf FIFO.

Adrian.

**Comment by antiprism** [comment]
Hi Mase

I forgot to mention, the binary package (which I am going to look at now) will install dependencies, which will include the i2c-tools package (although not a strict dependency) which contains i2cdetect.

These are the updated mpd.conf install/uninstall scripts
[mpd_oled_volumio_scripts2.zip](https://github.com/antiprism/mpd_oled/files/5936633/mpd_oled_volumio_scripts2.zip)

Sometimes, after running them, the first click of the play button doesn't work. They could be quietened a bit more. I haven't checked what happens depending if a non-MPD audio source is playing.

Adrian.

**Comment by supercrab** [comment]
Ah yes, I forgot about MPD being a service - should be easy to stop and start then.  

I couldn't get the install script to update the `mpd_oled` service parameters to run from volumio (because of permission errors) but I wonder if I can run each command separately to get around this.

Scripts look good, cheers!

**Comment by antiprism** [comment]
Hi Mase

I don't know about the permission errors. Normally the commands in a script run with the same user/permissions that you run the script.

I have an i2c SSH1106 attached (-o 6) and none of the other -o settings cause a segfault for me, but if the wrong model is chosen then I imagine the program could write to memory that it shouldn't and cause a segfault.

Adrian.

**Comment by supercrab** [comment]
I think it's to do with the `install.sh` needing `sudo` to work. The current user is `volumio`.  In my tests calling `/usr/bin/sudo install.sh` failed but killing processes with `/usr/bin/sudo /usr/bin/killall mped_oled`  works.  Maybe `install.sh` would work if I preceded each line in the script with `/usr/bin/sudo`?   I'll have a look when I finish the plugin...  I'm sure it's gonna be fun & games..

**Comment by supercrab** [comment]
I2C scanner finished,  `UIConfig.json`, install/remove scripts to go...

![0EB55A49-3269-4C45-AE1B-749AF6111ABE](https://user-images.githubusercontent.com/1100068/107123340-e9175100-6894-11eb-82bb-074cd5ee848f.jpeg)

**Comment by supercrab** [comment]
Hi @antiprism 

I've attached the install and uninstall scripts that the plugin will call for your you to review.  I haven't tested it from the plugin yet as I thought I'd get some feedback first.  It removes `cava` and `mpd_oled` during the uninstall, is that what you would expect?

[plugin-install.zip](https://github.com/antiprism/mpd_oled/files/5938769/plugin-install.zip)

Cheers
Mase

**Comment by misko903** [comment]
Hi guys, just an idea - I think there is a plugin already using display
integrated on HAT. Maybe they have some knowledge how to do it? The plugin
is called pirateaudio, but I think nanosound have it too... But maybe I am
late :)

Michal


Dňa ne 7. 2. 2021, 12:00 Mr Mase <notifications@github.com> napísal(a):

**Comment by supercrab** [comment]
@misko903 I had a look  at the pirateaudio hat and it looks really cool!  It has a colour screen and lots of stuff built onto the hat too.  I had a look at the plugin that configures the screen and it wasn't very configurable - it only had 2 options.  The MPD_OLED application that @antiprism is really configurable, can use different OLED screens and now it'll have a Volumio plugin to help set it up and configure it :)

My plugin should be ready soon :)

**Comment by antiprism** [comment]
Hi Mase

Good work on the i2c scanner.

The binary deb package I am making for mpd_oled will install the mpd_oled and mpd_oled_cava (cava) binaries, as well as the support scripts (like the ones that change /etc/mpd.conf) and other support files, as well as automatically installing the other package dependencies. It will also be possible to upgrade the package, and to uninstall the package (which will delete everything that was installed).

If you install the programs from source, the same files (the important ones anyway) will be installed, and to the same place as the package (if not by default then this can be configured). mpd_oled probably supports `sudo make uninstall`, and maybe cava too (but cava only installs two files anyway), which will (should) remove everything installed by `make install-strip`.

The package will be downloaded, and then installed by a command like `sudo apt install ./mpd_oled_0.2_01012021_amrfh.deb`
It will be configured by the user by running scripts like the ones that set up the service, or the ones that set up a FIFO.

The package scripts are ready now, and I just have to adapt the existing documentation. I may have a package ready in the next couple of days.

For the installation of the plugin, you could rely on the package manager for installing and removing mpd_oled (you could download or bundle the mpd_oled package) and then call the installed scripts from your scripts rather than using modified versions. This should make the plugin installation and removal easier to write, and also more robust.

Looking at your scripts and https://volumio.github.io/docs/Plugin_System/Writing_A_Plugin.html , it looks like the plugin install.sh is run as root, but maybe as volumio with sudo, so if you create files under /home/volumio it may be only root can delete them, which you may want. If you don't want to keep the directories afterwards you could work in a temprorary directory and delete it afterwards (example in the script). If anyone has looked at mpd_oled before but not actually installed it then the install script will fail to install mpd_oled. The uninstall doesn't uninstall mpd_oled from the system, but it might be better to leave it if a package manager is not involved.

If you need root privileges for settings but are not doing this from the install script and so the script is running as volumio, it is probably possible to make a small helper binary that will run as root and apply the settings: https://www.computerhope.com/jargon/s/setuid.htm . I would add this to the binary package.

Adrian.

**Comment by antiprism** [comment]
Hi Mase

I was originally going to implement configurable layouts first, and I have already written the difficult code for this, but it requires lots of other changes before producung a working OLED display. I decided instead to reverse the order and add the layouts last, so I can introduce changes incrementally.

The currently planned steps are:

1. Add the autotools build and make a binary that can work on all players (almost complete)
2. Switch to U8g2 for the OLED contoller and graphics
3. Add a few fixed layout variations (with and without spectrum) and support all common OLEDs
4. Add configurable layouts

U8g2 is monochrome only, but I'll look at whther colour displays can be supported, as I have one available.

Adrian.

**Comment by supercrab** [comment]
Yeah, it's a lot of work isn't it?  Essentially a rewrite to support all this stuff and implement U8g2. 

I've managed to get the plugin into a state of ready to test but for some reason the `plugin publish` command isn't working as expected and it's not publishing my changes to github. :(

**Comment by supercrab** [comment]
New volumio plugin is ready for testing!

1) Activate SSH on your system : SSH
2) Connect to your device via your console (linux) or Putty (Windows)
3) Download the plugin:
`wget https://github.com/balbuze/volumio-plugins/raw/master/plugins/audio_interface/volparametriceq/volparametriceq.zip`

4) Now create a temporary folder `mkdir ./mop`
5) Unzip the previously downloaded file miniunzip `mpd_oled.zip -d ./mop`
6) Go to this folder `cd ./mop`
7) Install the plugins `volumio plugin install` and wait - It will take several minutes to do the building, then:
8) Remove all temporary files : `cd .. rm -Rf mop`
9) In the webUI, go to installed plugins section and enable the plugin.
10) Go in the plugin settings, change Oled type, hit save!

Plugin install

Won't change mpd_oled folder if it already exists
Won't change cava folder if it already exists

Plugin uninstall

Removes mpd_oled folder
Removes cava folder
Remove mpd conf changes

**Comment by misko903** [comment]
So if I want to test it, I will leave everything as it is (display is
working now) and proceed with steps you wrote. Right?

Michal

Dňa po 8. 2. 2021, 22:54 Mr Mase <notifications@github.com> napísal(a):

**Comment by supercrab** [comment]
@misko903  yes, follow the install steps :D  CAVA and mpd_oled won't be changed so it should be a pretty quick install. :)

**Comment by misko903** [comment]
The link in step 3 is wrong. I will search for the right one.
Also the miniunzip word in step 5 should be styled as code

miso


ut 9. 2. 2021 o 0:03 Mr Mase <notifications@github.com> napísal(a):

**Comment by misko903** [comment]
sorry for spamming, I see, on GitHub it is already corrected. I red the
email only...
miso


ut 9. 2. 2021 o 0:16 miso <misko0903@gmail.com> napísal(a):

**Comment by supercrab** [comment]
Ah yes, I updated the page when I tested the steps.  No worries.

**Comment by misko903** [comment]
Excellent, sirs!
on my RPI3B+ works everything ok, as now. Do you want some debug info? How to obtain it? My config is exactly as descridbed on webpage, but I have 20 bars + Reversed display.

**Comment by supercrab** [comment]
Awesome!  Just give it a good test of the settings if you can.  You'l have to type `jounalctl -f` and it will list some debug info but if `mpd_oled` is running the log will get filled up quickly.  You might find it easier to disable the plugin, make your changes in settings and then press Save, you'll see what parameters mpd_oled is being called with.

What parameters did you use before with mpd_oled?

**Comment by misko903** [comment]
`./mpd_oled -o 6 -b 21 -g 1 -f 20 -R`

**Comment by supercrab** [comment]
@misko903 have you changed the settings in the plugin to match?

**Comment by misko903** [comment]
I love the settings - especially scrolling, as before the scrolling was too slow for me.

Just one thing about which i am thinking longer - OLED display is degraded after 2.5 years of using. Maybe adding a screen saver option will help to reduce the wear? Should I create a new ticket?

**Comment by misko903** [comment]
> @misko903 have you changed the settings in the plugin to match?

Yes, it is working also with my original` ./mpd_oled -o 6 -b 21 -g 1 -f 20 -R` settings

**Comment by supercrab** [comment]
There is an invert option and timeout period (in hours) that will invert the screen periodically. Maybe @antiprism could change it from hours to minutes sonic inverts quicker?

I don’t think these cheap little screens are designed to be open for a long time, so perhaps a solution could be to clear the screen if no music has played in a certain while?

**Comment by antiprism** [comment]
Hi Mase

Congratulations on getting plugin into an installable state! I'll take a look later today.

I have made a first binary package, and tested it on a copy of Volumio on a Pi Zero with a factory reset. It can be installed like this
```
# install package
sudo apt update
sudo apt install gdebi-core
(cd /tmp && wget http://antiprism.com/misc/mpd-oled_0.2+202102081602+ed17ee2_armhf.deb)
sudo gdebi -n /tmp/mpd-oled_0.2+202102081602+ed17ee2_armhf.deb

# does it work?
sudo mpd_oled_volumio_mpd_conf_install.sh
mpd_oled -o 6
```
Unfortunately, although 'apt' (or 'aptitude') should be able to install the remote package, it doesn't on Volumio. 'apt-get' should be able to install the local package, but doesn't on Volumio. 'gdebi' works, but isn't installed by default and is only for installing local package files. The commands therefore go the roundabout way of installing gdebi, downloading the mpd_oled package and then installing the downloaded mpd_oled package with gdebi.

The install also installs the dependencies. I added i2c-detect to recommended packages, but it did not install, so I will add it to dependencies.

The package is removed with
```
sudo apt remove -y mpd-oled
```
I will finish udating the documentation and then merge the development into the master branch.

Adrian.

**Comment by antiprism** [comment]
Hi Michal

I'll add a screen saver.

The scroll is configurable  on the command line with `mpd_oled -s`.

You can see all the options by running `mpd_oled -h`, but it would probably be helpful if I listed them in the documentation, so I'll do that.

Adrian.

**Comment by antiprism** [comment]
Hi Mase

The invert option will take a decimal number, e.g. `mpd_oled -I 0.001` inverts the display every 3 seconds or so. I'll document this if it is not clear. I used hours as I didn't think people would want to see the screen changing often, but that was just a guess.

Adrian.

**Comment by misko903** [comment]
maybe we can add it to settings on volumio GUI?
which file should be translated? I can start with Slovak, maybe Czech...

**Comment by supercrab** [comment]
@antiprism Cool, I didn't realise invert took a decimal, I thought it was just integers.  I'll add that to the user config, does 0.001 = 3.6 seconds?  I found this in the code `(period > 0) ? (fmod(time(0) / 3600.0, 2 * period) > period) : period;` but I wasn't sure what the intention is.  I'll try the latest binaries now..

**Comment by supercrab** [comment]
@misko903 I've attached a language file for translation :)  Thanks

[strings_sk.zip](https://github.com/antiprism/mpd_oled/files/5950948/strings_sk.zip)

**Comment by antiprism** [comment]
Hi Mase

The period can be - 0: always normal, -1: alway inverted, positive decimal: invert with that period. For fmod(time(0) / 3600.0, 2 * period), time(0) gives the current time in seconds (relative to some date) and then get the remainder when divided by twice the period.

So, if period is 0 or -1 then period is returned, and 0=false and -1=true (non-zero). For a positive period, the remainder is either less than the period=false, or greater than the period=true, o it is normal then inverted each for the length of the period.

Not especially clear! I'll comment it, but also, maybe it doesn't all have to be just one line...

Adrian.

**Comment by supercrab** [comment]
I've tested the parameters for it and seems to work fine.  So would a value of 0.001 refresh every 3.6 seconds?

**Comment by antiprism** [comment]
Hi Mase

-I 0.001 is a period of 3.6 seconds, but the display only changes on whole numbers of seconds, so the inverting isn't very regular at that frequency.

Adrian.

**Comment by supercrab** [comment]
I renamed the script as `install2.sh` but didn't get anything on the spectrum analyser, mpd.conf has the correct changes.

` bash install2.sh
{"time":1612873146004,"response":"stop Success"}Failed to stop mpd.service: Access denied
Warning: Stopping mpd.service, but it can still be activated by:
  mpd.socket
rm: cannot remove '/tmp/mpd_oled_fifo': No such file or directory
`

**Comment by supercrab** [comment]
i stopped it, changed the audio settings and it seems to be working now.

**Comment by supercrab** [comment]
volumio@volumiotest:~/mop$ sudo bash install2.sh
{"time":1612873970631,"response":"stop Success"}rm: cannot remove '/tmp/mpd_oled_fifo': No such file or directory
volumio@volumiotest:~/mop$ mpd_oled -o 3

**Comment by antiprism** [comment]
The error mssage the first time said "Access denied", probably because of not using sudo.

The other eror messages always occur. I will quieten them after a testing period.

If mpd_oled is running as a service then there may be issues with the order that the services run. I would think that MPD should be running before mpd_oled, so the install script might need to enforce that by waiting for MPD to be active.

I am going to rename the three service scripts mpd_oled_install.sh -> mpd_oled_service_install.sh, etc, for clarity.

**Comment by misko903** [comment]
> @misko903 I've attached a language file for translation :) If you want to to a Czech one then it's you just need to copy the file and change the filename to `string_cs.json`
> 
> [strings_sk.zip](https://github.com/antiprism/mpd_oled/files/5950948/strings_sk.zip)
> 
> **But I wouldn't do any translation just yet as I think I might have some more changes to come...**

Translated. Here you go!

[strings_sk.zip](https://github.com/antiprism/mpd_oled/files/5951611/strings_sk.zip)

**Comment by supercrab** [comment]
@antiprism I think the install script does work!  I've been using a dodgy radio station, that isn't outputting any audio :|  I don't have any audio device on this pi 0...   I can change the script to make it silent `-f`

@misko903  Thanks!  Will add this now!

**Comment by antiprism** [comment]
I have just checked, and the deb file, which I made on Volumio (as the older OS), also installs and works on Moode, I couldn't find a way to install the deb file from the url on Moode/buster., but apt will install the downloaded deb file and so gdebi is not needed there.

**Comment by supercrab** [comment]
Misko can you translate this JSON please?  It's the new settings for screen inverter.  Thanks!
`	"INVERT_DISPLAY_PERIOD": {
		"LABEL":  "Period",
		"DOC": "The frequency that the display is periodically inverted to help avoid screen burn",
		"DEFAULT": "Disabled (default)",
		"ITEM1": "1 minute",
		"ITEM2": "2 minutes",
		"ITEM3": "3 minutes",
		"ITEM4": "6 minutes",
		"ITEM5": "12 minutes",
		"ITEM6": "24 minutes",
		"ITEM7": "48 minutes",
		"ITEM8": "1 hour",
		"ITEM9": "2 hours",
		"ITEM10": "3 hours",
		"ITEM11": "6 hours"
	},`

**Comment by antiprism** [comment]
The `mpd_oled_edit.sh` script in the binary package has the wrong path for the executable. When it is run it should be edited to say `ExecStart=/usr/local/bin/mpd_oled... `

**Comment by supercrab** [comment]
@misko903 I've updated the zip of the plugin so, you should be able to uninstall it and reinstall it using the old instructions.

**Comment by supercrab** [comment]
> The `mpd_oled_edit.sh` script in the binary package has the wrong path for the executable. When it is run it should be edited to say `ExecStart=/usr/local/bin/mpd_oled... `

I don't need to change anything with the install scripts do I?

**Comment by antiprism** [comment]
It is actually in `mpd_oled_install.sh`. It means that running the service edit script with options won't work until the path has been set correctly by running the edit script without options. I'll make a new deb package later today. I am scripting the package creation from the outset, so it is just a question of changing the script to rewrite this path, and then generating a new package. It doesn't affect  the scripts that change mpd.conf on Volumio.

**Comment by misko903** [comment]
"INVERT_DISPLAY_PERIOD": { "LABEL": "Perióda", "DOC": "Frekvencia, ako často je invertovaný displej na spomalenie vypaľovania obrazovky", "DEFAULT": "Zakázané (prednastavené)", "ITEM1": "1 minúta", "ITEM2": "2 minúty", "ITEM3": "3 minúty", "ITEM4": "6 minút", "ITEM5": "12 minút", "ITEM6": "24 minút", "ITEM7": "48 minút", "ITEM8": "1 hodina", "ITEM9": "2 hodiny", "ITEM10": "3 hodiny", "ITEM11": "6 hodín" },

**Comment by supercrab** [comment]
@misko903 thanks for that I've updated the zip file again!

**Comment by supercrab** [comment]
@antiprism I'll keep the plugin installer as is with the tried and tested building of `CAVA` and `mpd_oled` for now :)

**Comment by antiprism** [comment]
Hi Mase

Ah, I have just seen that you aren't using the development code (some of my answers assumed that you were!). I will be updating the repository with this code soon (probably within the next few days), and the build is painfully slow on a Pi Zero, but you could just check out the current version for a while, until the binary package has had some testing.

Adrian.

**Comment by supercrab** [comment]
@antiprism Yeah, I wasn't sure what I was gonna do and then chose the option that I understood the best and that seemed to work at the time.  It should be easy to change when your happy with the dev version.

I've given up doing development on my Pi 0.  It's not too bad on the Pi 3 mind. 

Here's the plugin in my repository (for testing)

<https://github.com/supercrab/volumio-plugins/blob/master/plugins/miscellanea/mpd_oled/>

Link to Volumio community plugin announcement post.

<https://community.volumio.org/t/plugin-mpd-oled-installation-configuration-plugin/44823>

**Comment by antiprism** [comment]
Hi Mase

Good work! Well done!

I look at the Spanish translation for the plugin. It should be ready later today.

Adrian.

**Comment by supercrab** [comment]
Hi @antiprism 

Thanks!  I'm glad to get it sorted.  It took longer than I thought it would and there were some unforseen tricky parts.

Looking forward to the translation cheers!

Mase

**Comment by antiprism** [comment]
Hi Mase

Here is the Spanish translation: [strings_es.zip](https://github.com/antiprism/mpd_oled/files/5966737/strings_es.zip)

Adrian.

**Comment by supercrab** [comment]
@antiprism Brilliant!  I've updated the plugin and added both your names to the credits @misko903.

**Comment by Maschine2501** [comment]
Hey Guys! Awesome work!
If you want, i can do a Translation to German... ;)

**Comment by supercrab** [comment]
@Maschine2501 Greetings and thanks!
If you're able to, please do, that would be great! 🥇

**Comment by Maschine2501** [comment]
Ok, i'll do my best to get it ready til saturday ;)

**Comment by supercrab** [comment]
@Maschine2501 Much appreciated :)

**Comment by supercrab** [comment]
@Maschine2501 The German translation has been done by someone on the Volumio forums but you can review it here:

https://github.com/supercrab/volumio-plugins/blob/master/plugins/miscellanea/mpd_oled/i18n/strings_de.json

**Comment by misko903** [comment]
At least we have now all world important languages - english, slovak, spanish and german 😄

**Comment by supercrab** [comment]
Ha, yeah!  That's the most important languages covered, so I'll do a pull request soon I think.

It's great how everyone in the community has pulled together and all helped out! 😁   I'm surprised...

**Comment by supercrab** [comment]
@antiprism I think we can close this issue now :D  Oh, someone asked about changing the clock format, shall I wait for you to do that before knocking my plugin on the head?

**Comment by Maschine2501** [comment]
> 
> 
> @Maschine2501 The German translation has been done by someone on the Volumio forums but you can review it here:
> 
> https://github.com/supercrab/volumio-plugins/blob/master/plugins/miscellanea/mpd_oled/i18n/strings_de.json

I was to slow...
I have had a look at it: looks allright ;)

**Comment by antiprism** [comment]
Hi Mase

As Rick, who requested the date format option, is using the plugin I'll add it into the current mpd_oled code.

I'll leave a couple of days for any final follow-ups and then close the issue.

Adrian.

**Comment by misko903** [comment]
Is there any plan when the official version will be available on volumio plugins "shop"?

**Comment by supercrab** [comment]
@misko903 you'll have to ask the Volumio development team that - I have no idea.

@antiprism Let me know when you know what the parameters will be and I'll implement it 😃

**Comment by antiprism** [comment]
Hi Mase

I have just added in option -d, which sets the date format to MM-DD-YYYY.

If this change is working correctly then it will probably be the last update before I push the change with the new build system, In which case, when you are changing the plugin you will probably want to change the install script to either clone mpd_oled at the commit of this last update, or to get the mpd_oled source from another place. I will wait for you to make this change I push the new build system.

Adrian.

**Comment by misko903** [comment]
![16133805423831594722916802003975](https://user-images.githubusercontent.com/41118056/107927081-d11b8d80-6f76-11eb-805c-539a87691386.jpg)
Like this? I have it from beginning. 
What about to switch off the display completely after longer idle?

**Comment by antiprism** [comment]
Hi Michal

Your date format is the default DD-MM-YYYY, if you set option -d the date will be displayed as 02-15-2021

I will add in an option to blank the display when idle for a specified time, but this will be after I update with the new build system.

Adrian.

**Comment by supercrab** [comment]
Hi @antiprism 

Cool, I'll update my plugin now.
 
I want to plugin to be up to date with `mpd_oled`  so I'll change my install script to use the new build system.  :)

Cheers
Mase

**Comment by antiprism** [comment]
Hi Mase

I can imagine that I will make several updates that break the plugin. You could keep the plugin in working condition by using your own clone of the mpd_oled repository for the plugin. This will allow you to synchronise the plugin with the mpd_oled changes, because the mpd_oled changes will not be available to people using the plugin until you push them to your repository.

Adrian.

**Comment by supercrab** [comment]
Cool, I shall change my install script to use a fork of `mpd_oled` on my GitHub.

Let me know when `mpd_oled` is stable again please and I'll update the plugin :)

**Comment by supercrab** [comment]
Hi @antiprism 

Is it worth mentioning in the readme.md that the plugin is also available if people can configure `mpd_oled` from within Volumio or is that overstepping the mark?

https://github.com/supercrab/volumio-plugins/tree/master/plugins/miscellanea/mpd_oled

**Comment by antiprism** [comment]
Hi Mase

I was going to mention the plugin, but I thought it would be available in Volumio soon and so I would mention it then. Is it expected to be included soon? If not I will add a link to your project page.

Adrian.

**Comment by supercrab** [comment]
I made the pull request for the plugin to be included into Volumio 12 days ago but it hasn't been acted on yet.   I don't know if they intend to include it or not.  I have not heard anything from them about it yet either.  

I did a showcase of `mpd_oled` and the plugin on 2 Facebook Volumio groups and the Volumio community forum.  Everyone loved it and it got thoroughly tested.  More people added translations which was great too.

I'm hoping it will be included because lots more people are running `mpd_oled` since my marketing drive and a few of them asked about it.  I had to tell them I had no idea when inclusion would be. 

I will let you know of any updates of course :)

Mase

**Comment by supercrab** [comment]
The plugin will not be made official.  I've had my plugin request denied due to changes to mpd.conf :(

_Thanks for your work. Unfortunately, we can't publish it as an official plugin. The install script modify mpd.conf.tmpl in /Volumio. This will prevent Volumio to update.
2 solutions.
1 - use a workaround with loopback as it is done with plugin such as equalizer
2 (better) write your plugin for Volumio buster. A new beta should be available soon. We can help you. Have a look at peppymeter plugin. It works like that.
Sorry for that .._

**Comment by antiprism** [comment]
Hi Mase

I have updated the README to give a link to your plugin at Github

   https://github.com/antiprism/mpd_oled#volumio

The loopback solution made using the software mixer is not too difficult to implement and might be compatible with a number of setups. Here are the details

https://community.volumio.org/t/current-state-of-cava-visualizer-support-especially-for-streaming-sources/44976/8?u=adrii

It won't be for everyone (being a software mixer), but could be the default installation method for the plugin. Then, if people want they could just run the mpd_oled scripts that allow changing mpd.conf on the command line and configure the FIFO in the plugin, and this might not be an issue for the Volumio team if they are not obligatory scripts.

Adrian.

**Comment by supercrab** [comment]
Hi @antiprism 

Thanks for that, much appreciated!

Thanks for the pointers.  I tried using software mixing using the same ALSA configuration.  It took a while to get it right and once it did work, it caused the Spotify plugin to stutter.  I had to knock it on the head because of this.  Also, when you change the mixer to hardware and back to software you lose all the configuration settings.

The buster release of Volumio is still being developed and things are change.  One of the maintainers Balbuze couldn not get  `cava` to work with FIFO because the format is compatible.

Mase

**Comment by antiprism** [comment]
Hi Mase

I haven't felt like promoting the ALSA approach, as I don't want to deal with configuration issues, and so I haven't received a lot of feedback relating to it (not that there should be any problems in theory with the approach). It is interesting that you had issues like that.

I don't know what the format issue is with Cava, but Cava used to need a fixed format written to the FIFO, but the format can now be specified in the configuration file (sample_rate and sample_bits).

Adrian.

**Comment by supercrab** [comment]
@antiprism  On my own system, I gave up on the ALSA configuration and used a short cable looping from the output back into the mic input!  It works well though :)

I don't have Buster installed to give it any proper investigation at the moment, I'll let you know when I look into it.

Based on your suggestion: change the installer so it doesn't mess with mpd conf.  I could have a button on the plugin that changes the MPD config and another one to remove the changes.  This might allow me to get around the issue of the installer making these changes...

Mase

**Comment by antiprism** [comment]
Closing. The Volumio 3 plugin is now available.

**Comment by misko903** [comment]
yes, finally!
thank you!
miso


so 5. 11. 2022 o 7:53 Adrian Rossiter ***@***.***> napísal(a):

**Comment by supercrab** [comment]
I did it!  Although the fix was not something I would have found by myself 😬

**Comment by misko903** [comment]
Awesome, thanks!
miso


ne 6. 11. 2022 o 9:58 Mr Mase ***@***.***> napísal(a):
